### DISCUSSION TITLE
Install specific version of an AUR package

### DISCUSSION THREAD

**Discussion by arakis** [Morganamilo/paru#1068] [Q&A] [answered]
I want to know how to install a specific version and/or branch/commit/tag.

I know ArchLinux is a Rolling Distribution and partial Updates are not allowed. That's why i'm using archive.archlinux.org.

There are two reasons why I need to specify a specific version (or commit):

a) If the AUR packages require a SPECIFIC installed dependency of the regular packages. Example: You stick with `virtualbox 7.0.10` (using archive.archlinux.org), but the latest AUR ` virtualbox-ext-oracle` is at ` 7.0.12`. To work around this, instead of using paru, I'm using these commands:
```
cd ~/.cache/paru/clone/virtualbox-ext-oracle
git checkout 9184ed4abf39476502419f626e3da6980ccf53d2 #7.0.10
makepkg -si --noconfirm
```

b) avoid re-downloading packages/sources again. The reason is, once a month, I completely rebuild my arch linux from another running linux chroot via a large, well maintained script. This means, I wipe the whole file system except the arch package cache, and the paru cache and some important data directories. I can re-build as often I want, normally, there are no re-downloads. But when paru detects new version, it will force always the latest versions. In case of large binary packages, i there could be several gigabytes of data. FYI: I'm on cell connection only, and mostly I have horrible connection.

I have no problems to iterate the commits and check the PKGBUILD version until I find a specifc commit that matches that PKGBUILD file. But even if I have the matching commit, I have no way to tell paru to take that commit. Even if I checkout the commit manually (see my example above), paru will ignore this.

Of course, I could take makepkg alone without paru, but then I lose some dependency resolving features, like when an AUR packages requires another AUR package.

**Comment by Morganamilo** [comment]
You best bet is to store your pkgbuilds somewhere either via git or on disk then add a pkgbuild repo to paru with

```
[myreponame]
URL = https://github.com/foo/bar
# OR
Path= /foo/bar
```

Then you can have your pkgbuild do whatever you want

**Reply by Morganamilo** [reply]
As a less effort hack you could just change the version in paru's cache and commit that change. You'll get a conflict on updates though.

**Reply by Morganamilo** [reply]
The first option is better though as paru will take your version as the latest so not want to update.

**Reply by arakis** [reply]
I'm not sure if you got my problem correctly. In my opinition, there are two ways, i would prefer:

a) Specify specific version like `paru -S virtualbox-ext-oracle=7.0.10`. It seems, this is not supported.
b) Regardless of the download state/cache state, just take the cloned directory "as it is".

Let's say, the cloned (default) dir is `~/.cache/paru/clone/virtualbox-ext-oracle`. Now, I cd in, and git checkout the desired branch/commit. This is up to me. Now, I wish to tell paru, just take that directory as it is, without new online version check and so son. Like a `makepkg -si`, but with paru's dependency resolving features.

By the way, for the "checkout specific branch feature", for that I would write a bash script, that just take the latest commit before a specific date. With that in mind, you have an `archive.archlinux.org` like mechanism, and everything "should" be compatible to each other. But again, this is my part and has nothing todo with the my request for a) or b).

**Reply by Morganamilo** [reply, marked as answer]
> paru -S virtualbox-ext-oracle=7.0.10.

This won't ever happen because the AUR's API will only give us info on the current version of a package so all dependency resolving and sanity checks go out the window. So it wouldn't really make a sensible feature.

I still think the pkgbuild repo method I suggested is what you want though.

```
# create a pkgbuild repo and add it to paru.conf
morganamilo@starlight ~ % mkdir oldpkgs
morganamilo@starlight ~ % sudo tee -a /etc/paru.conf <<-
EOF

  [oldpkgs]
  Path = file:///home/morganamilo/oldpkgs
EOF

# take virtualbox-ext-oracle from the AUR and put it in the repo
morganamilo@starlight ~ % git -C oldpkgs clone https://aur.archlinux.org/virtualbox-ext-oracle
# check out to the version we want
morganamilo@starlight ~ % git -C oldpkgs/virtualbox-ext-oracle checkout 9184ed4abf39476502419f626e3da6980ccf53d2

# now paru sees  virtualbox-ext-oracle-7.0.10-1 and it will take priority over the AUR
morganamilo@starlight ~ % paru -Si virtualbox-ext-oracle
Repository      : oldpkgs
Name            : virtualbox-ext-oracle
Version         : 7.0.10-1
Description     : Oracle VM VirtualBox Extension Pack
URL             : https://www.virtualbox.org/
Groups          : None
Licenses        : custom:PUEL
Provides        : None
Depends On      : virtualbox=7.0.10
Make Deps       : None
Check Deps      : None
Optional Deps   : rdesktop: client to connect vm via RDP
Conflicts With  : None
Path            : /home/morganamilo/oldpkgs/virtualbox-ext-oracle

morganamilo@starlight ~ % paru -S virtualbox-ext-oracle
:: Resolving dependencies...
:: Calculating conflicts...
:: Calculating inner conflicts...

Pkgbuilds (1)                  Old Version  New Version    Make Only
oldpkgs/virtualbox-ext-oracle               7.0.10-1       No

:: Proceed with installation? [Y/n]:

**Reply by arakis** [reply]
Thank you for this detailed reply, now i understand your idea/suggestion better :-)

**Comment by jeinstei** [comment]
I am running into the same issue, especially as `pacman` supports the feature. I thought I'dd add a (much delayed) voice to this. Is this resolved in an issue somewhere?

Edit: https://github.com/Morganamilo/paru/issues/1084
